### ISSUE TITLE
Quantization Issue Yolov11/v10/v9 -- Inference time and memory usage deteriorated

### ISSUE THREAD

**Issue by widroit2025** [ultralytics/ultralytics#19446]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

Hi everyone,

I am working with the different releases of Yolo, i.e., v9, v10, and v11, in order to implement a custom model able to detect objects on the edge of a memory-constrained device.
For this reason, I am investigating the usage of quantization provided inside the Ultralytics library.
I tried several combinations of format and precision, specifically, mnn, openvino, tflite, etc with FP16 and INT8.
However, when the models are running in lower precision (ex:int8) the inference time and the memory usage increase with respect to FP32. 
The same behavior for FP16, namely if I cast the .pt into mnn, openvino, or other formats compatible with FP16 quantization, the inference time and memory usage experience a deterioration.
I tried to change the activation functions into something that is more quantization-friendly (ReLU) but with no results.
I have wandered around the issue section of the repo but once again without any success.
So, I am asking how is it possible such behavior and in case if there are some alternative solutions to effectively perform quantization.

Thank you in advance.

### Environment

Ultralytics 8.3.58 🚀 Python-3.12.7 torch-2.3.1+cu121 CUDA:0 (NVIDIA GeForce RTX 3060, 11940MiB)
Setup complete ✅ (8 CPUs, 31.2 GB RAM, 36.1/191.5 GB disk)

OS                  Linux-6.11.0-18-generic-x86_64-with-glibc2.40
Environment         Linux
Python              3.12.7
Install             pip
RAM                 31.22 GB
Disk                36.1/191.5 GB
CPU                 AMD Ryzen 3 PRO 4350G with Radeon Graphics
CPU count           8
GPU                 NVIDIA GeForce RTX 3060, 11940MiB
GPU count           1
CUDA                12.1

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.9.0>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.0>=1.4.1
torch               ✅ 2.3.1>=1.8.0
torch               ✅ 2.3.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.18.1>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.13>=2.0.0

### Minimal Reproducible Example

from ultralytics import YOLO

# Load the YOLO11 model
model = YOLO("yolo11n.pt")

# Export the model to TFLite format
model.export(format="tflite", int8=True) 

# Load the exported TFLite model
tflite_model = YOLO("yolo11n_int8.tflite")

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @widroit2025, thank you for your interest in Ultralytics 🚀! We appreciate you taking the time to provide detailed information about the quantization issue you're experiencing.

If this is a 🐛 Bug Report, please ensure that a minimum reproducible example (MRE) is included to help us debug the issue effectively. The script you shared already provides some valuable context. If you can further narrow it down or provide any additional observations, that would be helpful.

If this is a custom training ❓ Question, feel free to provide more details about your specific use case, such as changes in model architecture, dataset examples, or logs from training and inference.

As you're using quantization and experiencing deteriorations in memory usage and processing time, we recommend ensuring that the issue is not isolated to specific versions of the library or environment. You can upgrade your Ultralytics package and verify that all dependencies are up-to-date within a Python>=3.8 environment. 

YOLO models can be run in various environments like cloud-based notebooks, virtual machines, or using containerized Docker images, which are pre-equipped with dependencies. Ensuring that your setup is consistent with recommended environments can also assist in identifying potential discrepancies.

An Ultralytics engineer will review your issue and provide further assistance soon. Please stay tuned for updates. Thank you for bringing this to our attention! 🚀

**Comment by glenn-jocher** [comment]
@widroit2025 thank you for reporting this quantization performance issue. Based on your description, here are some key considerations:

1. **Hardware-Software Alignment**: The performance gains from quantization (FP16/INT8) are highly dependent on hardware acceleration support. For NVIDIA GPUs like your RTX 3060, we recommend using TensorRT exports with proper calibration as shown in the [TensorRT INT8 quantization guide](https://docs.ultralytics.com/integrations/tensorrt#exporting-tensorrt-with-int8-quantization). Ensure you're using the latest TensorRT version and proper workspace allocation.

2. **Calibration Data**: INT8 quantization requires representative calibration data. Verify your `data` argument points to a valid dataset YAML during export, as shown in the [TFLite export documentation](https://docs.ultralytics.com/integrations/tflite#usage).

3. **Edge Device Considerations**: For memory-constrained devices, we recommend:
   - Using TensorRT with `dynamic=True` and proper batch sizing
   - Validating quantization-aware training (QAT) approaches
   - Reviewing our [model deployment best practices](https://docs.ultralytics.com/guides/model-deployment-practices)

The Ultralytics team is actively working to improve quantization performance across formats. Please share specific benchmark comparisons (FP32 vs INT8/FP16) for your target hardware if this issue persists after verifying these parameters.

**Comment by TheAnswer96** [comment]
Thank you Glenn for reaching me.
I am relying on quantization built-in procedure inside Ultralytics library.
The quantized model currently runs on Raspberry Pi 5 as well we tested also on 3B+.
We performed calibration feeding more than 300 images, and more specifically a subset of relevant instances.
During the quantization we did not experience any warning/errors at all, suggesting that the procedure completed smoothly.

For the performance benchmark I attach below some instances.

**MODEL PERFORMANCE COMPARISON - v10n**

OPENVINO
---------------------------------------
FP32:
  Memory: 694.31 MB (max 868.69)
  CPU: 81.87% (max 95.10%)
  Power: 7.17W (max 11.19W)
  Processing Time: 91.72ms

FP16:
  Memory: 684.51 MB (max 858.75)
  CPU: 81.35% (max 94.90%)
  Power: 7.18W (max 10.98W)
  Processing Time: 91.58ms

INT8:
  Memory: 733.44 MB (max 889.25)
  CPU: 85.87% (max 97.60%)
  Power: 7.16W (max 11.08W)
  Processing Time: 302.00ms

MNN
---------------------------------------
FP32:
  Memory: 596.86 MB (max 758.00)
  CPU: 46.23% (max 57.10%)
  Power: 6.24W (max 8.39W)
  Processing Time: 120.39ms

FP16:
  Memory: 595.33 MB (max 759.06)
  CPU: 46.02% (max 57.10%)
  Power: 6.23W (max 8.37W)
  Processing Time: 121.18ms

INT8:
  Memory: 601.22 MB (max 759.55)
  CPU: 46.27% (max 55.80%)
  Power: 6.25W (max 8.25W)
  Processing Time: 120.68ms

TFLITE
---------------------------------------
FP32:
  Memory: 1182.13 MB (max 1404.91)
  CPU: 80.10% (max 100.00%)
  Power: 7.17W (max 12.82W)
  Processing Time: 307.00ms

FP16:
  Memory: 1169.26 MB (max 1368.25)
  CPU: 79.82% (max 100.00%)
  Power: 7.19W (max 12.62W)
  Processing Time: 308.98ms

Note: Processing Time = preprocessing + inference + postprocessing

The code used to profile those data is the following:

def profile_function(func, *args, **kwargs):
    """Profiles memory and CPU usage of a function execution."""
    cpu_usage = []
    stop_event = threading.Event()

    def monitor_cpu():
        while not stop_event.is_set():
            cpu_usage.append(psutil.cpu_percent(interval=0.1))

    cpu_thread = threading.Thread(target=monitor_cpu)
    cpu_thread.start()

    # Force garbage collection before measurement
    gc.collect()

    result = None

    # Define a wrapper to capture the function result
    def wrapper():
        nonlocal result
        result = func(*args, **kwargs)

    # Measure memory while running the function
    mem_usage = memory_usage(wrapper)

    # Stop CPU monitoring thread
    stop_event.set()
    cpu_thread.join()

    # Calculate statistics
    avg_mem_usage = sum(mem_usage) / len(mem_usage)
    max_mem_usage = max(mem_usage)
    avg_cpu_usage = sum(cpu_usage) / len(cpu_usage) if cpu_usage else 0
    max_cpu_usage = max(cpu_usage) if cpu_usage else 0

    return result, avg_mem_usage, max_mem_usage, avg_cpu_usage, max_cpu_usage

In addition to that, we noticed that after quantization the number of layers with type FP32 is the same with some additional quant and dequant layers. While quant and dequant layers are reasonable in some points, it appears to me really strange this behavior.

Hope to have some support on my activities.
Thank you in advance.

**Comment by glenn-jocher** [comment]
@TheAnswer96 thank you for sharing detailed benchmarks. For Raspberry Pi deployments, we recommend focusing on TensorFlow Lite with XNNPACK enabled for ARM optimization, as this often yields better quantization benefits than OpenVINO/MNN on ARM devices. The observed performance regression could stem from: 

1. Lack of hardware-accelerated INT8 ops in your current runtime (quant/dequant layers add overhead without compute benefits)
2. Suboptimal operator fusion in exported graphs
3. Memory alignment issues in ARM NEON implementations

You might achieve better results by:
- Compiling TFLite with XNNPACK enabled
- Using `--half` instead of INT8 for ARMv8 devices
- Exploring NVIDIA Jetson platforms with native TensorRT support

The Ultralytics team is working to improve ARM quantization support in future releases. For immediate needs, consider profiling individual model layers to identify bottlenecks using our [benchmark utility](https://docs.ultralytics.com/reference/utils/benchmarks/).

**Comment by TheAnswer96** [comment]
Hi Glenn

I followed Ultralytics guidelines on RPi deployment but there is no mention to XNNPACK.
In addition, the issue I mentioned arises when I export (using the built-in function) the model regardless the deployment on RPi or Desktop, so Intel environment.
Concerning what you suggest of relying on TensorRT, it may work but unfortunately I am not interested to edge GPU execution.

Can you give further details on how to achieve a working quantization leveraging XNNPACK? Looking around the repo, I failed to find any official guide. Even external pointer would be appreciated.

Anyway, I would like to remark that export with half=True or Int8=True failed.
I attached below an example of the resultant Neural network architecture.

![Image](https://github.com/user-attachments/assets/a7a70ed1-f96a-475b-b170-8daf3a177358)

Please, guide me in solving the issue I am facing...
Thank you in advance.

**Comment by glenn-jocher** [comment]
Thank you for the detailed follow-up. For TFLite quantization, while Ultralytics doesn't directly expose XNNPACK controls, you can enable it post-export by compiling TFLite with XNNPACK flags and setting `interpreter.set_num_threads()` + `XNNPack` delegate in your inference code. Our current TFLite export focuses on baseline quantization - we'll enhance ARM optimizations in future releases. 

The mixed precision layers in your export suggest partial operator support. For full INT8 quantization:
1. Use `model.export(..., int8=True, data="your_dataset.yaml")` with representative data
2. Verify all ops support quantization via [TFLite docs](https://www.tensorflow.org/lite/performance/quantization_spec)

We're tracking quantization improvements at #12544 and welcome specific hardware/use-case details to prioritize enhancements.

**Comment by TheAnswer96** [comment]
Thank you for the answer.
It seems that XNNPack can be the solution for TFLITE.
I will let you know whether it will work.
Nonetheless the strange behavior I mentioned in my very first request is related to MNN and OpenVino as well.
Moreover, regardless the machine (Intel CPU, RTX 3060, ARMv8) , executing those models returned extremely uneffiencient performance for INT8 and FP16.
Is this a known limitation of ultralytics library? if not how can I solve this issue?

Best!

**Comment by glenn-jocher** [comment]
@TheAnswer96 thank you for the detailed benchmarks. Quantization performance depends heavily on hardware acceleration support, which varies by format/device. For OpenVINO INT8, we recommend using the [OpenVINO Optimization Guide](https://docs.openvino.ai/) to ensure proper configuration. The team is actively working on quantization improvements tracked in GitHub Issue #12544 - please share your specific hardware/config details there to help prioritize enhancements. 

For MNN/ARM deployments, ensure you're using the latest MNN library with NEON optimizations enabled. We observe best INT8 results on devices with native quantized op support (e.g., Snapdragon NPUs).

**Comment by TheAnswer96** [comment]
Thank you for the answer.
What I was trying to say, it's that neural network after quantization appears broken.
In other words, the quantized model contains dummy layers (as shown in the previous image I attached), and so it is not a matter of hardware acceleration. 
I would like to know if the architecture shown in the image I attached is anomalous or not.
Thank you in advance.

**Comment by glenn-jocher** [comment]
The quant/dequant layers you observed are expected in quantized models and part of standard TFLite/OpenVINO quantization workflows. For Ultralytics models, these layers should be optimized during export when hardware acceleration is properly configured. If you're seeing performance degradation despite this, please:  
1. Verify you're using the latest `ultralytics>=8.3.60`  
2. Share your exact export command and calibration dataset YAML  
3. File a [GitHub Issue](https://github.com/ultralytics/ultralytics/issues) with your model architecture visualization and benchmark comparisons.  

We'll investigate if there's a framework-specific optimization opportunity in our export pipeline. For technical details on quantization layers, see the [TFLite quantization documentation](https://www.tensorflow.org/lite/performance/quantization_spec).

**Comment by TheAnswer96** [comment]
Hi Glenn

Concerning point 1), I have the latest ultralytics release.
to export the model I relied on the following function based on export ultralytics command.

```
def convert2desired(m_path, d_path, format, type="FP32"):
    model = YOLO(m_path)
    if type == "FP32":
        model.export(format=format, data=d_path, batch=1, task="detect")
    elif type == "FP16":
        model.export(format=format, data=d_path, half=True, batch=1, task="detect")
    else:
        model.export(format=format, data=d_path, int8=True, batch=1, task="detect")
    return
```
the parameter format assumes the following values ["tflite", "openvino", "mnn"].
I have used the format expressively supported in the documentation.

The yaml file is the following:
```
train: learning\\train\\images
val: learning\\val\\images
test: learning\\test\\images

names:
    0: object_0
    1: object_1
    2: object_2
    3: object_3
    4: object_4
    5: object_5
```

let me know whether what I did was correct or not.

**Comment by glenn-jocher** [comment]
Thank you for sharing your export configuration. Your code follows valid syntax, but there's an important consideration with your dataset YAML file. The `data` parameter expects paths to **annotation files** (like `dataset.yaml` with image paths and labels), not direct image directories. For proper quantization calibration, ensure your YAML points to a valid dataset configuration as shown in the [Dataset Configuration Guide](https://docs.ultralytics.com/datasets/detect/). 

If quantization issues persist, please share:
1. Full error traceback
2. Validation results before/after export
3. Complete environment details via `yolo settings`

We'll help investigate further.
